STM32Cube/Core/Src/my2c.c

```diff
@@ -0,0 +1,144 @@
+#include <stdbool.h>
+#include "my2c.h"
+
+#include "stm32h7xx_hal.h"
+#include "stm32h7xx_hal_i2c.h"
+
+#include "FreeRTOS.h"
+#include "semphr.h"
+
+SemaphoreHandle_t I2CBinarySemaphore;
+
+extern I2C_HandleTypeDef hi2c4;
+
+uint8_t timeout_MS = 200;
+
+/**
+ * Register i2c4 callbacks, semaphore
+ */
+bool MY2C_init(void)
+{
+    if (HAL_I2C_RegisterCallback(&hi2c4, HAL_I2C_MEM_TX_COMPLETE_CB_ID, &I2C4_MemTxCallback) != HAL_OK)
+    {
+        return false;
+    }
+
+    if (HAL_I2C_RegisterCallback(&hi2c4, HAL_I2C_MEM_RX_COMPLETE_CB_ID, &I2C4_MemRxCallback) != HAL_OK)
+    {
+        return false;
+    }
+
+    if (HAL_I2C_RegisterCallback(&hi2c4, HAL_I2C_ERROR_CB_ID, I2C4_ErrorCallback) != HAL_OK)
+    {
+        return false;
+    }
+
+	I2CBinarySemaphore = xSemaphoreCreateBinary();
+
+	if (I2CBinarySemaphore)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * Internal helper read function, read N bytes using interrupt handling
+ */
+HAL_StatusTypeDef MY2C_readNByteRegister_IT(uint8_t address, uint8_t reg, uint8_t *data, uint8_t len)
+{
+	HAL_StatusTypeDef status = HAL_OK;
+
+	status = HAL_I2C_IsDeviceReady(&hi2c4, address << 1, 50, 500);
```

Comment:
Am I correct that this blocks for up to 500ms?

Reply:
Removed isDeviceReady in favour of using the returned status from the HAL read/write methods; isDeviceReady is blocking and redundant since the read/write checks for ready status.